snuba/optimize_tracker.py

```diff
@@ -0,0 +1,81 @@
+from abc import ABC, abstractmethod
+from datetime import datetime
+from typing import Optional, Set
+
+from snuba import state
+from snuba.state import rds
+
+
+class OptimizedPartitionTracker(ABC):
+    """
+    OptimizedPartitionTracker is an abstract class for tracking progress of an optimize job.
+    """
+
+    @abstractmethod
+    def get_completed_partitions(self) -> Optional[Set[str]]:
+        """
+        Get a list of partitions that have completed optimization.
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def update_completed_partitions(self, part_name: str) -> None:
+        """
+        Add partitions that have completed optimization.
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def remove_all_partitions(self) -> None:
+        """
+        Remove all partitions that have completed optimization.
+        """
+        raise NotImplementedError
+
+
+class InMemoryPartitionTracker(OptimizedPartitionTracker):
+    """
+    Tracks progress of partitions that have completed optimization using in memory data storage.
+    """
+
+    def __init__(self) -> None:
+        self.__tracker: Set[str] = set()
+
+    def get_completed_partitions(self) -> Optional[Set[str]]:
+        return self.__tracker if len(self.__tracker) > 0 else None
+
+    def update_completed_partitions(self, part_name: str) -> None:
+        self.__tracker.add(part_name)
+
+    def remove_all_partitions(self) -> None:
+        self.__tracker.clear()
+
+
+class RedisOptimizedPartitionTracker(OptimizedPartitionTracker):
+    """
+    This class keeps track of partitions which have already been optimized by keeping state of
+    optimized partitions in redis.
+    """
+
+    def __init__(self, host: str, expire_time: datetime) -> None:
+        self.__bucket = f"{state.optimize_prefix}{host}"
```

Comment:
If you're going to have the prefix be configurable, it should be passed into the `__init__`, otherwise, you're making an implicit tie between this class and the `state` module, and it's not clear why

Reply:
I would also say that the redis client should be passed into this as opposed to implicitly assumed to be the `snuba.state.rds` object.

Reply:
The `expire_time` should be part of the bucket as well. That way you're not relying on the expiration mechanism to separate the optimize jobs

Reply:
> If you're going to have the prefix be configurable, it should be passed into the `__init__`, otherwise, you're making an implicit tie between this class and the `state` module, and it's not clear why

The prefix in now locally defined in the file. I had incorrectly assumed that all redis related definition should be in state.

> I would also say that the redis client should be passed into this as opposed to implicitly assumed to be the snuba.state.rds object.

Done

> The expire_time should be part of the bucket as well. That way you're not relying on the expiration mechanism to separate the optimize jobs

While it's not harmful to add the expire time to the bucket, I would not like to do so since it could imply that optimize jobs can run for more than a day when someone looks at a bucket name like `snuba-optimize:some-hostname:2022-04-01 23:00:00`.

Reply:
> While it's not harmful to add the expire time to the bucket, I would not like to do so since it could imply that optimize jobs can run for more than a day 

I understand that they are not *supposed to* run for more than a day but let's say that the mechanism we have to enforce fails for some reason. Would we rather have these keys collide or not collide? If we would rather have them collide that's fine by me

Reply:
Added the date to the time bucket to avoid chances of conflict